cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/ZabbixWidget.kt

```diff
@@ -0,0 +1,223 @@
+        }
+    }
+
+    private fun fetchAuthorizationToken() {
+        val body = attachCredentialsToBody()
+        httpPost(publicUrl, body)
+    }
+
+    private fun attachCredentialsToBody(): JsonObject {
+        val credentials = JsonObject()
+                .put(CogboardConstants.PROP_USER, user)
+                .put(CogboardConstants.PROP_PASSWORD, password)
+        return prepareRequestBody(USER_LOGIN, credentials)
+    }
+
+    override fun handleResponse(responseBody: JsonObject) {
+        val result = extractResult(responseBody)
+        when {
+            hasResponseData(result) -> sendUpdate(result)
+            isAuthorized(result) -> fetchItemData()
+            hasError(responseBody) -> sendAuthorizationError(responseBody)
+            else -> {
+                saveToken(result.toString())
+                fetchItemData()
+            }
+        }
+    }
+
+    private fun extractResult(responseBody: JsonObject): Any {
+        val body = responseBody.getString(BODY)
+        val value = JsonObject(body)
+        return value.getValue(RESULT, "")
+    }
+
+    private fun hasResponseData(result: Any) = result.toString().contains("key_")
+
+    private fun sendUpdate(result: Any) {
+        val responseParams = result as JsonArray
+        val lastValue = responseParams.extractValue(LAST_VALUE)
+        send(JsonObject()
+                .put(LAST_VALUE, lastValue)
+                .put(NAME, responseParams.extractValue(NAME))
+                .put(HISTORY, modifyHistory(responseParams))
+                .put(CogboardConstants.PROP_WIDGET_STATUS, getStatusResponse(lastValue)))
+    }
+
+    private fun modifyHistory(responseParams: JsonArray): JsonObject {
+        val history = fetchHistoryFromContent()
+        val lastUpdate = responseParams.extractValue(LAST_CLOCK).toLong()
+        val filteredHistory = history.filter { it.key.toLong() > lastUpdate.minus(HOUR_IN_MILLS) }
+        val lastValue = responseParams.extractValue(LAST_VALUE)
+        return JsonObject(filteredHistory).mergeIn(JsonObject().put(lastUpdate.toString(), lastValue), true)
+    }
+
+    private fun fetchHistoryFromContent(): Map<String, Any> {
+        val content = boardServ.getContent(config.getString(CogboardConstants.PROP_ID))
+        return content.getJsonObject(HISTORY, JsonObject()).map
+    }
+
+    private fun getStatusResponse(lastValue: String): Widget.Status {
+        val convertedValue = lastValue.toDouble()
+        return when {
+            metricHasMaxValue() -> status(convertToPercentage(convertedValue))
+            metricHasProgress() -> status(convertedValue)
+            else -> Widget.Status.UNKNOWN
+        }
+    }
+
+    private fun metricHasMaxValue() = metricHasProgress() &&
+            METRICS_WITH_MAX_VALUE.contains(selectedMetric)
+
+    private fun metricHasProgress() = METRICS_WITH_PROGRESS.contains(selectedMetric)
+
+    private fun convertToPercentage(convertedValue: Double): Double {
```

Comment:
so this will convert any `double` to a percentage `double` value? `1.1` -> `110.0`%?
Let me know

Reply:
And this method should be tested by JUnit